alibi/prototypes/protoselect.py

```diff
@@ -0,0 +1,543 @@
+                             n_splits: int = 2,
+                             batch_size: int = int(1e10),
+                             preprocess_fn: Optional[Callable[[np.ndarray], np.ndarray]] = None,
+                             **kwargs) -> dict:
+    """
+    Cross-validation parameter selection for ProtoSelect with Euclidean distance. The method computes
+    the best epsilon radius.
+
+    Parameters
+    ----------
+    refset
+        Tuple, `(X_ref, X_ref_labels)`, consisting of the reference data instances with the corresponding reference
+        labels.
+    protoset
+        Tuple, `(X_proto, )`, consisting of the prototypes selection set. Note that the argument is passed as a tuple
+        with a single element for consistency reasons.
+    valset
+        Optional tuple `(X_val, X_val_labels)` consisting of validation data instances with the corresponding
+        validation labels. 1-KNN classifier is evaluated on the validation dataset to obtain the best epsilon radius.
+        In case ``valset=None``, then `n-splits` cross-validation is performed on the `refset`.
+    num_prototypes
+        The number of prototypes to be selected.
+    eps_range
+        Optional ranges of values to select the epsilon radius from. If not specified, the search range is
+        automatically proposed based on the inter-distances between `X_ref` and `X_proto`. The distances are filtered
+        by considering only values in between the `quantiles` values. The minimum and maximum distance values are
+        used to define the range of values to search the epsilon radius. The interval is discretized in `grid_size`
+        equal-distant bins.
+    quantiles
+        Quantiles, `(q_min, q_max)`, to be used to filter the range of values of the epsilon radius. See `eps_range` for
+        more details. If not specified, no filtering is applied. Only used if ``eps_range=None``.
+    grid_size
+        The number of equal-distant bins to be used to discretize the `eps_range` proposed interval. Only used if
+        ``eps_range=None``.
+    n_splits
+        The number of cross-validation splits to be used. Default value 2. Only used if ``valset=None``.
+    batch_size
+        Batch size to be used for kernel matrix computation.
+    preprocess_fn
+        Preprocessing function to be applied to the data instance before applying the kernel.
+
+    Returns
+    -------
+    Dictionary containing
+     - ``'best_eps'``: ``float`` - best epsilon radius according to the accuracy of a 1-KNN classifier.
+     - ``'meta'``: ``dict`` - dictionary containing argument and data gather throughout cross-validation.
+    """
+    X_ref, X_ref_labels = refset
+    X_proto = protoset[0]
+    if preprocess_fn is not None:
+        X_ref = _batch_preprocessing(X=X_ref, preprocess_fn=preprocess_fn, batch_size=batch_size)
+        X_proto = _batch_preprocessing(X=X_proto, preprocess_fn=preprocess_fn, batch_size=batch_size)
```

Comment:
Curious why we have to do it in batches?

Reply:
One can use large models (e.g. transformers) to extract a feature representation, for which you need to do the preprocessing in batches.